js/modules/types/attachment.js

```diff
@@ -0,0 +1,99 @@
+const MIME = require('./mime');
+const { arrayBufferToBlob, blobToArrayBuffer, dataURLToBlob } = require('blob-util');
+const { autoOrientImage } = require('../auto_orient_image');
+
+// Increment this everytime we change how attachments are upgraded. This allows us to
+// retroactively upgrade existing attachments. As we add more upgrade steps, we could
+// design a pipeline that does this incrementally, e.g. from version 0 (unknown) -> 1,
+// 1 --> 2, etc., similar to how we do database migrations:
+const CURRENT_PROCESS_VERSION = 1;
+
+// Schema version history
+//
+// Version 1
+//   - Auto-orient JPEG attachments using EXIF `Orientation` data
+//   - Add `schemaVersion` property
+
+// // Incoming message attachment fields
+// {
+//   id: string
+//   contentType: MIMEType
+//   data: ArrayBuffer
+//   digest: ArrayBuffer
+//   fileName: string
+//   flags: null
+//   key: ArrayBuffer
+//   size: integer
+//   thumbnail: ArrayBuffer
+//   schemaVersion: integer
+// }
+
+// // Outgoing message attachment fields
+// {
+//   contentType: MIMEType
+//   data: ArrayBuffer
+//   fileName: string
+//   size: integer
+//   schemaVersion: integer
+// }
+
+// Middleware
+// type UpgradeStep = Attachment -> Promise Attachment
+
+// UpgradeStep -> SchemaVersion -> UpgradeStep
+const setSchemaVersion = (next, schemaVersion) => async (attachment) => {
+  const isAlreadyUpgraded = attachment.schemaVersion >= schemaVersion;
+  if (isAlreadyUpgraded) {
+    return attachment;
+  }
+
+  let upgradedAttachment;
+  try {
+    upgradedAttachment = await next(attachment);
+  } catch (error) {
+    console.error('Attachment.setSchemaVersion: error:', error);
+    upgradedAttachment = null;
+  }
+
+  const hasSuccessfullyUpgraded = upgradedAttachment !== null;
+  if (!hasSuccessfullyUpgraded) {
+    return attachment;
+  }
+
+  // TODO: Enable `...` object spread operator syntax:
```

Comment:
I don't think we need this TODO. It's kind of an implicit 'go replace all instances of `Object.assign()`' isn't it?